[Feature] Gitlab automated setup

ivanchimeno opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

**Note:** This pr is based on [the Jenkins configuration-as-code](https://github.com/ls1intum/Artemis/pull/3080) branch so it might be easier to review when that pr is merged.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Setting up Gitlab and getting it to run with Artemis is time consuming due to the many manual steps needed (generate access tokens, disable prometheus monitoring, etc).

### Description
<!-- Describe your changes in detail -->
This pr attempts to simplify the set up process by:

- adding `docker/gitlab/gitlab-config.rb` file which includes some default configuration but can be freely modified by the user. This file configures Gitlab after the container is started.
- adding `docker/gitlab/gitlab-local-setup.sh` script which automatically generates the access tokens required by Artemis. It also enables outbound connections to local network.
- extending the `docker/gitlab-jenkins-mysql.yml` to include the added files from above
- adding a quick-start section in the documentation 
  
### Steps for Testing
The best way to test is to follow the documentation steps and see Gitlab is configured correctly.

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->
No change

dfuchss commented on src/main/docker/gitlab-jenkins-mysql.yml:
```diff
@@ -1,11 +1,14 @@
             - artemis-gitlab-config:/etc/gitlab
+            - ./gitlab/gitlab-local-setup.sh:/gitlab-local-setup.sh
+        environment:
+            - GITLAB_OMNIBUS_CONFIG="./gitlab/gitlab-config.rb"
```
How does this work? Do you map gitlab-config.rb somewhere to the container?

dfuchss replied:
See also comments for src/main/docker/gitlab/gitlab-config.rb

dfuchss commented on src/main/docker/gitlab/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM gitlab/gitlab-ce:latest
+
```
You may want to add the modifications regarding the password reset "bug" mentioned in the documentation?

dfuchss commented on src/main/docker/gitlab/gitlab-config.rb:
```diff
@@ -0,0 +1,75 @@
+## GitLab configuration settings
```
Instead of providing the whole file you may want to add something like that to the docker-compose environment:

```
 environment:
      - GITLAB_OMNIBUS_CONFIG=prometheus_monitoring['enable'] = false; nginx['listen_port'] = 80; gitlab_rails['monitoring_whitelist'] = ['0.0.0.0/0']

```

See also https://github.com/kit-sdq/Artemis-SAML2-Test-Docker/blob/main/docker-compose.yml#L29

ivanchimeno replied:
Ah that's interesting. I think the reason why I opted for a file, is so that the user can easily extend it in case he/she wants to add more options. Do you think the file is a bit of an overkill (I like your suggestion since we avoid creating an extra file. ?

dfuchss replied:
In addition, you won't get problems with newly added configuration values. If you provide a file, you have to make sure that this file is up to date according to the latest version of gitlab .. forever :) . If you do not provide a file. You have no need to update it again and again.

ivanchimeno replied:
That's a valid point. I will definitely change that!

dfuchss commented on src/main/docker/gitlab-jenkins-mysql.yml:
```diff
@@ -1,11 +1,14 @@
 version: "2"
```
I think we should use at least docker-compose version 3 :)

dfuchss reviewed (changes requested):
Generally, I think this guide looks good 👍 

I've just left some remarks :)

ivanchimeno commented:
> Generally, I think this guide looks good +1
> 
> I've just left some remarks :)

Thanks for the review @dfuchss ! I applied your suggestions :)

dfuchss commented on docs/dev/setup/jenkins-gitlab.rst:
```diff
@@ -124,8 +124,46 @@ Starting the Artemis server should now succeed.
+
+   ::
+
+        docker exec -it gitlab /bin/sh -c "sh gitlab-local-setup.sh"
```
I'm not sure whether the working dir of gitlab is "/" . I would suggest to use the full path .. `sh /gitlab-local-setup,sh`

dfuchss commented on src/main/docker/gitlab/gitlab-local-setup.sh:
```diff
@@ -0,0 +1,55 @@
+ADMIN_USER_ID="1"
+
+# Enter the url of the Gitlab instance.
+GITLAB__API_URL="http://localhost/api/v4/"
```
I would separate the values the user has to set and the values 'calculated'

So maybe just arrange ADMIN_PERSONAL_ACCESS_TOKEN and GITLAB__API_URL to the top. I don't think the other values have to be changed by default. (Especially, I think you have no need to set ADMIN_USER_ID AUTHORIZATION_BEARER_HEADER manually)

dfuchss replied:
Are the two underscores in GITLAB__API_URL that what you want? I think that could lead to errors.

dfuchss commented on src/main/docker/gitlab/gitlab-local-setup.sh:
```diff
@@ -0,0 +1,55 @@
+"
+
+echo
+echo 'The access Jenkins token has been created and can be copied into youur jenkins-casc-config.yml file:'
```
typo in "youur" (actually, I don't know how to edit such stuff inplace on github :) )

dfuchss reviewed (changes requested):
Just some small recommendations :)

ivanchimeno commented:
> Just some small recommendations :)

Thanks again :)

dfuchss reviewed (approved):
Nice work! Code looks comprehensible.

b-fein commented on docs/dev/setup/jenkins-gitlab.rst:
```diff
@@ -124,8 +124,46 @@ Starting the Artemis server should now succeed.
+
+   ::
+
+        docker-compose -f src/main/docker/gitlab-jenkins-mysql.yml up --build
```
```suggestion
        docker-compose -f src/main/docker/gitlab-jenkins-mysql.yml up --build -d
```
Detaching here would be nice, as then the same terminal can just be used after step 5 to terminate them as commented there.

b-fein commented on docs/dev/setup/jenkins-gitlab.rst:
```diff
@@ -124,8 +124,46 @@ Starting the Artemis server should now succeed.
+   ::
+
+        docker exec -it gitlab /bin/sh -c "sh /gitlab-local-setup.sh"
+
```
Here the other token can already be inserted into the `jenkins-casc-config.yml`. This should probably be mentioned.
Also the docker containers should probably be shut down here with `docker-compose -f src/main/docker/gitlab-jenkins-mysql.yml down` as the Jenkins Dockerfile is modified in the other section.

b-fein reviewed (changes requested):
I created a fresh setup following the documentation, builds trigger and cloning works.

It therefore works as intended, but I left comments with some changes to improve the documentation and an additional one I cannot make inline as it is not in the changed parts:

In the section Automated Jenkins Server Setup step 9
```
version-control:
    url: http://172.33.0.2:8081
```
~~the url should have either the port 80, or better be `http://localhost:8081` if the other PR with the fix for that is merged soon anyway (then also the Jenkins-URL can be adapted at the same time).~~
The URL here and the one for Jenkins should be `http://localhost:$port` (unless it has just been changed by the other merged PR).

ivanchimeno commented on docs/dev/setup/jenkins-gitlab.rst:
```diff
@@ -124,8 +124,46 @@ Starting the Artemis server should now succeed.
+
+   ::
+
+        docker exec -it gitlab /bin/sh -c "sh /gitlab-local-setup.sh"
```
Huh, I didn't that was possible. Should I remove `giltab` in your suggestion?

b-fein replied:
The `gitlab after `exec` has to stay, as otherwise docker-compose won’t know the service. Is there another one I’m not seeing at the moment?

ivanchimeno replied:
Sounds good, I just wasn't sure

ivanchimeno commented:
> I created a fresh setup following the documentation, builds trigger and cloning works.
> 
> It therefore works as intended, but I left comments with some changes to improve the documentation and an additional one I cannot make inline as it is not in the changed parts:
> 
> In the section Automated Jenkins Server Setup step 9
> 
> ```
> version-control:
>     url: http://172.33.0.2:8081
> ```
> 
> ~the url should have either the port 80, or better be `http://localhost:8081` if the other PR with the fix for that is merged soon anyway (then also the Jenkins-URL can be adapted at the same time).~
> The URL here and the one for Jenkins should be `http://localhost:$port` (unless it has just been changed by the other merged PR).

Thanks for the remark! I modified the url and added a new step telling the user to uncomment the lines for `internal-urls.ci-url` and `internal-urls.vcs-url`.

b-fein reviewed (approved):
Looks good!

sleiss reviewed (approved):
Changes look fine to me.
